### PR TITLE
Backup on MySQL DB slave if it is available

### LINKED ISSUE #218: Port eybackup_slave recipe
V4 version: https://github.com/engineyard/ey-cloud-recipes/tree/master/cookbooks/eybackup_slave

### PR DESCRIPTION
**Pull request by kaspergrubbe**
This fixes #218 .

# Forever cron

I also think that the old code included a "race"-condition.

If `node.dna['backup_window']` was at some point positive, and then turned off, the old cronjob was never deleted.

```ruby
if node.dna['backup_window'] != 0 && ['db_master','solo'].include?(node.dna['instance_role'])
  encryption_command = @encryption_command
  backup_command = "eybackup -e mysql #{encryption_command} >> /var/log/eybackup.log 2>&1"

  cron "mysql" do
    command backup_command
    month   '*'
    weekday '*'
    day     '*'
    hour    node['backup_hour']     # this attribute is set by ey-base/attributes/snapshot_and_backup_intervals.rb
    minute  node['backup_minute']
  end
end
```

# Postgresql

I also simplified the Postgresql code.

# Request for comments

Let me know what you think.

### CODE REVIEW

File: cookbooks/ey-backup/recipes/mysql.rb [hunk 1]
```diff
@@ -21,35 +21,45 @@
...
 end
 
-if node.dna['backup_window'] != 0 && ['db_master','solo'].include?(node.dna['instance_role'])
+has_backups_enabled  = node.dna['backup_window'] != 0
+db_slaves_available  = node.dna['db_slaves'].any?
+is_db_master_or_solo = ['db_master','solo'].include?(node.dna['instance_role'])
+is_db_slave          = 'db_slave' == node.dna['instance_role']
+is_first_slave       = node.dna['db_slaves'].first == (node['ec2'] && node['ec2']['local_hostname'] ? node['ec2']['local_hostname'] : hostname.stdout)
+
+if has_backups_enabled && (db_slaves_available && is_db_slave && is_first_slave || !db_slaves_available && is_db_master_or_solo)
```
Comment (johndalton) [inline cookbooks/ey-backup/recipes/mysql.rb]: `is_db_slave` is redundant in this test - we can just check:
```
if has_backups_enabled && (db_slaves_available && is_first_slave || !db_slaves_available && is_db_master_or_solo)
```

Comment (kaspergrubbe) [inline cookbooks/ey-backup/recipes/mysql.rb]: You are right!

Comment (kaspergrubbe) [inline cookbooks/ey-backup/recipes/mysql.rb]: This has been changed.

**Comment by mage2k** [comment]
Closing this as we're not going to automatically move backups to MySQL db_slaves as that could adversely affect many existing environments.  Once we get our current back log of other Pull Requests merged and released we will be making where the backup runs configurable via an attribute in either the ey-backup or mysql cookbook.

**Pull request closed without merging**